core: add dynamic managed storage dispatcher

sp-viktori opened the pull request:
Add an optional `adapter` field in the managed storage driver definition which will redirect all managed storage requests to a third party executable. This allows third party vendors to implement a direct storage interface instead of going through OpenStack/Cinder.

A similar dispatch mechanism is done for Vdsm in a separate PR.

## Changes introduced with this PR

* Add an optional `adapter` field in the managed storage `driver_info` map.

* If a managed storage has the new "adapter" field, instead of running the `cinderlib-client.py` helper, the `CinderlibExecutor` will instead direct all requests to an `{adapter}-adapter` helper. This is intended to be a vendor-provided executable implementing the same protocol as the default Cinderlib client.

## Are you the owner of the code you are sending in, or do you have permission of the owner?

Yes.

dupondje reviewed (changes requested):
Nice work!

Looks good. Except, can you split the CinderlibExector part and the dynamic managed storage part into 2 commits (can be on the same PR).
Just so clearly see what is refactoring, and what is additional code.

sp-viktori committed:
core: fix .gitignore missing /repository

Add missing /repository to the default .gitignore (present in
.gitignore.in).

Signed-off-by: Viktor Ivanov <viktor.ivanov@storpool.com>

sp-viktori committed:
core: refactor CinderLib* to ManagedBlock*

Refactor "CinderlibExecutor" classes to a more generic
"ManagedBlockExecutor".

Rename the log directory previously only meant to hold the CinderLib log
file to a more generic "managedblock" log directory.

Rename the data directory previously only meant to hold the CinderLib
client to a more generic "managedblock" directory. The existing
CinderLib client remains `cinderlib-client.py.in` as well as its log
file name and etc subdirectory containing strictly CinderLib-related
files.

Signed-off-by: Viktor Ivanov <viktor.ivanov@storpool.com>

sp-viktori committed:
core: add dynamic managed storage dispatcher

Add an optional "adapter" field in the managed storage driver definition
which will redirect all managed storage requests to a third party
executable.

Signed-off-by: Viktor Ivanov <viktor.ivanov@storpool.com>

sp-viktori committed:
core: refactor CinderStorageDao to ManagedBlock*

Refactor CinderStorageDao and related classes to ManagedBlockStorageDao.

Rename the cinder_storage table in the engine database to mb_storage,
including related stored procedures and SQL migration scripts.

Signed-off-by: Viktor Ivanov <viktor.ivanov@storpool.com>

sp-viktori committed:
packaging/setup: rename cinderlib DB to managedblock DB

Rename the CinderLib database to a more generic Managed Block database.

Rename the cinderlib plugin in engine-setup to "managedblock" including
all cinderlib engine config options.

Add a migration plugin in engine-setup, "cinderlib_rename", that detects
the now-legacy cinderlib options and renames them to their new
managedblock equivalents.

Signed-off-by: Viktor Ivanov <viktor.ivanov@storpool.com>

sp-viktori committed:
packaging/engine-backup: rename cinderlib to managedblock

Rename cinderlib DB to managedblock DB in the engine-backup script.

Keep cinderlib detection and restoring in engine-backup so a backup from
an older version can be correctly restored and later migrated by
engine-setup.

Signed-off-by: Viktor Ivanov <viktor.ivanov@storpool.com>

sp-viktori commented on backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/common/utils/managedblock/ManagedBlockExecutor.java:
This ended up being a bit of a rabbit hole, but it's not as bad as it looks 🤞 

There are still a couple of things to figure out and test properly:
1. Config (Setup) migration from all the `*CINDERLIB*` vars to the new `*MANAGEDBLOCK*` vars. I think if I add a setup step somewhere in one of the early stages I can detect the old cinderlib confs and migrate things (there is likely no need to be too smart about it and try to rename the cinderlib db itself).
2. engine backup and restore - I think if I fix point 1 above this will work out of the box, but need to test it. If I understand how things work, this is absolutely required for a hosted engine upgrade since it always must restore from backup?

If you have any hints about best way to proceed with either point, it will be much appreciated 🙏

sp-viktori replied:
Thank you! I'll check `aaaupgrade.py` and push again once I have the restore/migration stuff ready for another review. Might take a while to test things 🤞

sp-viktori committed:
core: support non-SCSI managed block attachments

Do not populate GUID metadata for BLOCK type attachments if they don't
have the `scsi_wwn` field populated.

Signed-off-by: Viktor Ivanov <viktor.ivanov@storpool.com>

dupondje reviewed (changes requested):
Some very minor comment. But otherwise very nice :)

dupondje commented on ovirt-engine.spec.in:
```diff
@@ -628,14 +628,15 @@ Requires:	makeself >= 2.2
-Summary:	Setup and upgrade specific plugins for cinderlib integration database
+%package setup-plugin-managedblock
+Summary:	Setup and upgrade specific plugins for managed block integration database
+Obsoletes:	%{name}-setup-plugin-cinderlib
```
@sp-viktori : Thanks
I think we better use
`Obsoletes:	%{name}-setup-plugin-cinderlib < 4.5.8`

As the current <= 4.5.7 will for example not obsolete version 4.5.7-1.

Next to that, the Provides is not needed. I think we just want to fix the packages that would still depend on the old package name, than to reply on the old package name supplied via provides.

sp-viktori replied:
Yep that would've likely be broken, good point.

Pushed an update to set it to `< 4.5.8` without `Provides:`.

dupondje replied:
Sigh :) OST failed because
```
E            Problem: cannot install both ovirt-engine-setup-plugin-ovirt-engine-4.5.8-0.master.20260402091157.git8a7f15e83fa.el9.noarch from extra-src-1 and ovirt-engine-setup-plugin-ovirt-engine-4.5.8-0.master.20260326124410.git81eb8559f9.el9.noarch from @System
E             - package ovirt-engine-setup-plugin-cinderlib-4.5.8-0.master.20260326124410.git81eb8559f9.el9.noarch from @System requires ovirt-engine-setup-plugin-ovirt-engine = 4.5.8-0.master.20260326124410.git81eb8559f9.el9, but none of the providers can be installed
E             - cannot install the best update candidate for package ovirt-engine-setup-plugin-ovirt-engine-4.5.8-0.master.20260326124410.git81eb8559f9.el9.noarch
E             - problem with installed package ovirt-engine-setup-plugin-cinderlib-4.5.8-0.master.20260326124410.git81eb8559f9.el9.noarchnon-zero return code
```

I think we better use
` Obsoletes:	%{name}-setup-plugin-cinderlib < 4.5.8-1`

Then %{name}-setup-plugin-cinderlib < 4.5.8-0.xxxx stuff is also obsoleted.

sp-viktori replied:
I'll get an OST up and running locally as well to avoid these in the future.

Meanwhile pushed an update to make it `4.5.8-1`.

dupondje commented:
/ost

dupondje commented:
/ost

dupondje commented:
/ost

gorsing commented:
When will the new version with this feature be released?
